dali/pipeline/operators/reader/loader/loader.h

```diff
@@ -102,65 +110,112 @@ class Loader {
       "Please overload PrepareEmpty for custom LoadTarget type other than Tensor");
   }
 
-
   // Get a random read sample
-  LoadTargetPtr ReadOne() {
+  LoadTargetSharedPtr ReadOne(bool is_new_epoch) {
     if (!loading_flag_) {
       PrepareMetadata();
     }
     TimeRange tr("[Loader] ReadOne", TimeRange::kGreen1);
-    // perform an iniital buffer fill if it hasn't already happened
+    // perform an inital buffer fill if it hasn't already happened
     if (!initial_buffer_filled_) {
       TimeRange tr("[Loader] Filling initial buffer", TimeRange::kBlue1);
+      shards_.push_back({0, 0});
 
       // Read an initial number of samples to fill our
       // sample buffer
       for (int i = 0; i < initial_buffer_fill_; ++i) {
-        auto tensor_ptr = LoadTargetPtr(new LoadTarget());
+        auto tensor_ptr = LoadTargetUniquePtr(new LoadTarget());
         PrepareEmpty(*tensor_ptr);
         ReadSample(*tensor_ptr);
+        ++read_sample_counter_;
+        if (IsNextShardRelative(read_sample_counter_ - 1, virtual_shard_id_)) {
+          samples_from_next_epoch_ += 1;
+          Index curr_elms = shards_.back().end;
+          shards_.push_back({curr_elms, curr_elms});
+        }
         sample_buffer_.push_back(std::move(tensor_ptr));
+        ++shards_.back().end;
       }
 
       // need some entries in the empty_tensors_ list
       TimeRange tr2("[Loader] Filling empty list", TimeRange::kOrange);
       std::lock_guard<std::mutex> lock(empty_tensors_mutex_);
       for (int i = 0; i < initial_empty_size_; ++i) {
-        auto tensor_ptr = LoadTargetPtr(new LoadTarget());
+        auto tensor_ptr = LoadTargetUniquePtr(new LoadTarget());
         PrepareEmpty(*tensor_ptr);
         empty_tensors_.push_back(std::move(tensor_ptr));
       }
 
       initial_buffer_filled_ = true;
     }
-    // choose the random index
-    int idx = shuffle_ ? dis(e_) % sample_buffer_.size() : 0;
 
```

Comment:
Please update the PR description with what major changes were done in the loader so that we can have some context while reviewing

Reply:
Updated, but I don't know if it is that easy to describe in the PR description.